File: Part 2 - Regression/Section 6 - Polynomial Regression/PLR.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Nov 29 11:24:33 2018

@author: giacomolini
"""

#Polynomial Linear Regression


#importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Import a dataset
dataset = pd.read_csv('Position_Salaries.csv') #name of the dataset
X = dataset.iloc[:, 1:-1].values 
Y = dataset.iloc[:, 2].values #modify extrems if necessary

#splitting into training and test datasets
"""from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2, random_state =0)"""
#possibly different test_size


#feature scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_Y = StandardScaler()
Y_train = sc_Y.fit_transform(Y_train)
"""
#usiamo due modelli per fare comparazione

#Fitting Linear Regression to data set
from sklearn.linear_model import LinearRegression
Lin_reg = LinearRegression()
Lin_reg.fit(X,Y)


#Fitting Polynomial Regression to data set
#trasformo prima la matrice originale in una matrice con le potenze
from sklearn.preprocessing import PolynomialFeatures
Poly_reg = PolynomialFeatures(degree=4)
X_poly = Poly_reg.fit_transform(X)
#faccio una MLR sulla matrice nuova
Lin_reg2 = LinearRegression()
Lin_reg2.fit(X_poly, Y)

#Visualising SLR Results
plt.scatter(X, Y, color = 'red')
plt.plot(X, Lin_reg.predict(X), color = 'blue')
plt.title('Prediciton with SLR')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()


#Visualising PLR Results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X, Y, color = 'red')
plt.plot(X_grid, Lin_reg2.predict(Poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Prediciton with PLR')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

#Prediction of result with SLR
Lin_reg.predict(6.5)


#Prediction of result with PLR
Lin_reg2.predict(Poly_reg.fit_transform(6.5))

File: Part 2 - Regression/Section 6 - Polynomial Regression/PLR.r
#Polynomial Linear Regression (PLR)

#import the dataset
#setwd("percorso_alla_cartella")
dataset = read.csv('Position_Salaries.csv') #name of the data set
dataset = dataset[, 2:3] # extrems of the coloumn i need

#splitting the data set into training and test
#install.packages('caTools')
#library(caTools)
#set.seed(123)
#split = sample.split(dataset$Purchased, SplitRatio = 0.8)#possibily different ratio
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

#feature scaling
# training_set[,2:3] = scale(training_set[,2:3])
# test_set[,2:3] = scale(test_set[,2:3])

# Fitting Simple Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ Level,
               data = dataset)

#Predicting result with SLR
Y_pred = predict(lin_reg,data.frame(Level=6.5))

# Visualising the SLR results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Salary vs Position_Level (SLR)') +
  xlab('Level') +
  ylab('Salary')

#Fitting Polynomial Regression Model to dataset
dataset$Level2 = (dataset$Level)^2
dataset$Level3 = (dataset$Level)^3
dataset$Level4 = (dataset$Level)^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

# Predicting Result with PLR
Y_pred = predict(poly_reg,data.frame(Level=6.5,
                                     Level2= 6.5^2,
                                     Level3= 6.5^3,
                                     Level4= 6.5^4))

# Visualising the SLR results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Salary vs Position_Level (PLR)') +
  xlab('Level') +
  ylab('Salary')